diff --git a/src/options_parser.c b/src/options_parser.c
index 26fe23a..39e0741 100644
--- a/src/options_parser.c
+++ b/src/options_parser.c
@@ -40,10 +40,10 @@ int read_query(struct Options* options) {
     const char EOL = '\n';
 
     // check if there is input
-    if ( (c = getchar()) == EOF ) return 0;
+    if ( (c = getchar()) == EOF || c == EOL) return 0;
     options->query[i] = c;
     i++;
-
+    
     while ( (c = getchar()) != EOL ) {
         options->query[i] = c;
         i++;
diff --git a/src/search.c b/src/search.c
index 157a3bc..fff7142 100644
--- a/src/search.c
+++ b/src/search.c
@@ -161,7 +161,7 @@ void search_query_datamap(struct Datamap* database, char query[], struct linked_
     struct Entry* current_entry;
    
     int begin = (query_size-3)<1?1:(query_size-3);
-    int end = (query_size+3)>database->size?database->size:(query_size+3);
+    int end = (query_size+3)>database->size?database->size-1:(query_size+3);
     for (int i = begin; i <= end; i++) {
         current = database->map[i]->start;
         if (current != NULL) {
